framework/src/base/Assembly.C

```diff
@@ -2092,6 +2136,39 @@ Assembly::reinitLowerDElemRef(const Elem * elem,
   }
 }
 
+void
+Assembly::reinitLowerDElemDualRef(const Elem * elem,
+                                  const std::vector<Point> * const pts,
+                                  const std::vector<Real> * const weights)
+{
+  mooseAssert(pts->size(),
+              "Currently reinitialization of lower d elements is only supported with custom "
+              "quadrature points; there is no fall-back quadrature rule. Consequently make sure "
+              "you never try to use JxW coming from a fe_lower object unless you are also passing "
+              "a weights argument");
+
+  _current_lower_d_elem = elem;
+
+  unsigned int elem_dim = elem->dim();
+
+  for (const auto & it : _fe_lower[elem_dim])
+  {
+    FEBase * fe_lower = it.second;
+    FEType fe_type = it.first;
+    FEShapeData * fesd = _fe_shape_data_dual_lower[fe_type];
+
+    fe_lower->set_calculate_dual(true);
```

Comment:
> You don't necessarily have a non-NULL `_fe_shape_data_dual_lower` for every `_fe_lower` do you? An `_fe_lower` is created anytime you call _either_ `buildLowerDDualFE` _or_ `buildLowerDFE`, but an `_fe_shape_data_dual_lower` is only created when you call `buildLowerDDualFE`.

That is correct. The current implementation only gives non-NULL `_fe_shape_data_dual_lower` only when the user choose to use the dual method.



> I ran into a segmentation fault the other day when trying to use a mixed-order discretization with `u/order=SECOND` and `lm/order=FIRST` and `lm/use_dual=true`.

That is right. The dual method requires `lm/order = u/order ` `use_dual=true`, otherwise there would be an issue. Where do you suggest I may add a warning about this, just in case someone uses different order for LM and u?

Reply:
Do we even need to use this API at all? If I had noticed this in the libMesh PR, I probably would've asked for it to be removed. There's no precedent for it in the FE hierarchy. We should know whether to `calculate_dual` based on the prerequests in `buildLowerDDualFE`. This is how I think a single `reinitLowerDElemRef` method should look:
```c++
void
Assembly::reinitLowerDElemRef(const Elem * elem,
                              const std::vector<Point> * const pts,
                              const std::vector<Real> * const weights)
{
  mooseAssert(pts->size(),
              "Currently reinitialization of lower d elements is only supported with custom "
              "quadrature points; there is no fall-back quadrature rule. Consequently make sure "
              "you never try to use JxW coming from a fe_lower object unless you are also passing "
              "a weights argument");

  _current_lower_d_elem = elem;

  unsigned int elem_dim = elem->dim();
  for (const auto & it : _fe_lower[elem_dim])
  {
    FEBase * fe_lower = it.second;
    FEType fe_type = it.first;

    fe_lower->reinit(elem, pts, weights);

    if (FEShapeData * fesd = _fe_shape_data_lower[fe_type])
    {
      fesd->_phi.shallowCopy(const_cast<std::vector<std::vector<Real>> &>(fe_lower->get_phi()));
      fesd->_grad_phi.shallowCopy(
                                  const_cast<std::vector<std::vector<RealGradient>> &>(fe_lower->get_dphi()));
      if (_need_second_derivative_neighbor.find(fe_type) != _need_second_derivative_neighbor.end())
        fesd->_second_phi.shallowCopy(
                                      const_cast<std::vector<std::vector<TensorValue<Real>>> &>(fe_lower->get_d2phi()));
    }

    if  (FEShapeData * fesd = _fe_shape_data_dual_lower[fe_type])
    {
      fesd->_phi.shallowCopy(const_cast<std::vector<std::vector<Real>> &>(fe_lower->get_dual_phi()));
      fesd->_grad_phi.shallowCopy(
                                  const_cast<std::vector<std::vector<RealGradient>> &>(fe_lower->get_dual_dphi()));
      if (_need_second_derivative_neighbor.find(fe_type) != _need_second_derivative_neighbor.end())
        fesd->_second_phi.shallowCopy(
                                      const_cast<std::vector<std::vector<TensorValue<Real>>> &>(fe_lower->get_dual_d2phi()));
    }
  }
}
```

Reply:
My concern was whether to reinitialize both dual and non-dual shape functions, as I imagine only one of them will be used. If yes, then yeah I agree this would be a better way to combine both functions and remove this API

Reply:
You need to compute the non-dual shape functions anyways in order to compute the dual shape functions. If you have a completely non-dual simulation then `get_dual_phi` will never get called, and `fe_lower->reinit` won't do any dual related computation. I don't see any scenario in the code I posted where you will be wasting work, but maybe I'm missing something...

Reply:
I just checked after making the changes. You are right: for a completely non-dual simulation, `get_dual_phi` is never called. So no additional computation will be caused due to the change. Very helpful comments! Appreciate it!